Core: Enable doc values by default, when appropriate

rjernst opened the pull request:
Doc values significantly reduced heap usage, which results in faster GCs. This change makes the default for doc values dynamic: any field that is indexed but not analyzed now has doc values. This only affects fields on indexes created with 2.0+.

Notes:
* Doc values for geo was already messed up.  With lat/lon subfields enabled, doc values settings were never passed through.  I have explicitly disabled doc values for geo points for now.
* Percolator queries with fielddata execution are broken (see comment in test).  We should just remove execution mode...

closes #8312

rmuir commented on src/main/java/org/elasticsearch/index/mapper/core/AbstractFieldMapper.java:
```diff
@@ -701,7 +717,7 @@ protected void doXContentBody(XContentBuilder builder, boolean includeDefaults,
             builder.field("store", fieldType.stored());
         }
-        if (includeDefaults || hasDocValues() != Defaults.DOC_VALUES) {
+        if (includeDefaults || hasDocValues() != defaultDocValues()) {
```
I'm not sure this logic is correct... fieldType is mutable right? We should compute a final boolean in the ctor if we need this instead.

rjernst replied:
fieldType is not mutable here.  It is frozen in the ctor of AbstractFieldMapper.  This needs to be overridable by subclasses so e.g. geo mapper can use false instead of basing on the field type.

rmuir commented:
> This also simplifies the way to set doc values by removing the fielddata format path (again, for 2.0+ indexes).

I'm concerned this change will disable docvalues for more users than it will enable it for. Please, if we are going to remove this "way of doing it" then the system must fail hard if it sees it.

rjernst commented:
I changed the `return null` in case of `fielddata.format=doc_values` to an IAE.  IndexFieldDataService already has other places it checks values and throws IAE, so this should be fine I think.

clintongormley commented:
Would be nice get this syntax change https://github.com/elastic/elasticsearch/issues/8693#issuecomment-84784051 done at the same time (but as a separate PR, so it doesn't block this one)

rjernst commented:
>> I'm concerned this change will disable docvalues for more users than it will enable it for. Please, if we are going to remove this "way of doing it" then the system must fail hard if it sees it.

> I changed the return null in case of fielddata.format=doc_values to an IAE. IndexFieldDataService already has other places it checks values and throws IAE, so this should be fine I think.

Actually, this produced a lot of test failures.  There are still numerous tests using this old way of specifying doc values.  I've added back the old way, to simplify this change.  We should still simplify doc values to have only one way to enable/disable, but this can be done in another PR.

rjernst commented:
`_timestamp` was also giving problems with tests, randomly causing various test failures. The failures seemed to happen whenever _timestamp was enabled, and doc values explicitly disabled.  Any subsequent updates that did not contain _timestamp would cause a merge conflict.  I've set _timestamp back to doc values disabled by default.  We can improve this in a subsequent issue.

nirmalc commented:
Currently if the field length is more than 32766 , doc_values throws an error ( we avoid ignore_above analyzer in analyzed fields for same ) , Any helpers there planned ?

rjernst commented:
@nirmalc This is an internal limitation of the format used for doc values. I do wonder why you have an unanalyzed value that long, but in any case, you will still be able to disable doc values explicitly, and any full text that is analyzed won't have doc values enabled (it can't).

nirmalc commented:
As you guessed it its bad data from feeds we get mostly,  i guess we'll end up adding validation in data ingestion side as doc_Values would be great addition there  in all other cases.

rjernst commented:
@s1monw @jpountz any thoughts here?

@clintongormley In #8312 you had two extra ideas about disabling fielddata by default for `analyzed` strings and adding a token limit for `not_analyzed` strings. I think we should move these to separate issues? They are good ideas, but at least the latter I know will require a little thought in the mappers because the setting default becomes dynamic, which is similar to what made doc values by default a little complex here.

jpountz commented on src/main/java/org/elasticsearch/index/mapper/core/AbstractFieldMapper.java:
```diff
@@ -339,16 +341,22 @@ protected AbstractFieldMapper(Names names, float boost, FieldType fieldType, Boo
+        } else if (fieldDataType != null && FieldDataType.DOC_VALUES_FORMAT_VALUE.equals(fieldDataType.getFormat(indexSettings))) {
+            // convoluted way to enable doc values, should be removed in the future
+            this.docValues = true;
+         }else if (Version.indexCreated(indexSettings).onOrAfter(Version.V_2_0_0)) {
```
minor nitpick: there seems to be an indentation issue with the closing bracket

jpountz commented on src/main/java/org/elasticsearch/index/mapper/core/AbstractFieldMapper.java:
```diff
@@ -360,6 +368,14 @@ protected String defaultPostingFormat() {
     }
+    
+    protected boolean defaultDocValues() {
+        if (writePre2xSettings) {
```
Can we rename this variable to something like `indexCreatedBefore20` since it is the way it is used here if I understand correctly?

rjernst replied:
It used to be named something like that, but someone didn't like the "20" because they coudln't tell if it was 0.20 or 2.0. So maybe `indexCreatedBefore2x`?

jpountz replied:
sounds good to me!

jpountz commented on src/main/java/org/elasticsearch/index/mapper/core/AbstractFieldMapper.java:
```diff
@@ -360,6 +368,14 @@ protected String defaultPostingFormat() {
+    
+    protected boolean defaultDocValues() {
+        if (writePre2xSettings) {
+            return Defaults.DOC_VALUES;
```
Can we change the constant name to make it explicit that it only applies to 1.x indices?

jpountz commented on src/main/java/org/elasticsearch/index/mapper/core/AbstractFieldMapper.java:
```diff
@@ -360,6 +368,14 @@ protected String defaultPostingFormat() {
         return null;
     }
+    
+    protected boolean defaultDocValues() {
```
Should we make this method either abstract or return false to avoid traps? For instance I was worried that CompletionFieldMapper does not override it to return false. Maybe it works since completion fields are tokenized so the default value is false but I think making it explicit would be less error-prone?

rjernst replied:
I didn't want to make it abstract b/c the vast majority of mappers want the behavior here from the abstract mapper? But I'm fine adding it to completion field mapper explicitly.

rjernst commented:
@jpountz I pushed changes to address your comments.

rjernst commented:
I pushed another change which no longer allows overriding `hasDocValues()`. This forces subclasses who want to always disable doc values to explicitly pass `false` to the super ctor. It also simplifies the API, so there is only 1 method regarding doc values that can be overridden.

It is also worthwhile to note that with this PR, `_uid` will have doc values enabled (because it is indexed and not analyzed, although maybe we should be explicit there and always pass `true` to the super ctor?).

martijnvg commented on src/test/java/org/elasticsearch/timestamp/SimpleTimestampTests.java:
```diff
@@ -36,6 +37,7 @@
 
 /**
  */
+@Seed(value = "C50CB1C9767DAD4F")
```
I think this should be removed?

rjernst replied:
Oops!! Removed.

jpountz commented on src/main/java/org/elasticsearch/index/mapper/internal/AllFieldMapper.java:
```diff
@@ -160,9 +181,6 @@ protected AllFieldMapper(String name, FieldType fieldType, NamedAnalyzer indexAn
                              @Nullable Settings fieldDataSettings, Settings indexSettings) {
         super(new Names(name, name, name, name), 1.0f, fieldType, null, indexAnalyzer, searchAnalyzer,
```
Should it pass 'false' as 'docValues' too?

jpountz commented on src/main/java/org/elasticsearch/index/mapper/internal/TimestampFieldMapper.java:
```diff
@@ -229,6 +229,11 @@ protected TimestampFieldMapper(FieldType fieldType, Boolean docValues, EnabledAt
+    /*@Override
+    public boolean defaultDocValues() {
+        return false;
+    }*/
```
is it left on purpose?

rjernst replied:
Nope this was a mistake. Removed!

jpountz commented on src/main/java/org/elasticsearch/index/mapper/internal/VersionFieldMapper.java:
```diff
@@ -96,7 +96,7 @@ protected Field initialValue() {
 
     public VersionFieldMapper(Settings indexSettings) {
-        super(new Names(NAME, NAME, NAME, NAME), Defaults.BOOST, Defaults.FIELD_TYPE, null, null, null, null, null, null, indexSettings);
+        super(new Names(NAME, NAME, NAME, NAME), Defaults.BOOST, Defaults.FIELD_TYPE, false, null, null, null, null, null, indexSettings);
```
Should it be `true` instead?

jpountz replied:
Which makes me wonder that we should probably a test that the default mapping serializes to an empty json to make sure that this default logic works correctly?

rjernst replied:
I don't think that would have helped in this case, since `toXContent()` here just returns the builder. But yes, it should be `true` and it is now fixed.

rmuir commented:
> It is also worthwhile to note that with this PR, _uid will have doc values enabled (because it is indexed and not analyzed, although maybe we should be explicit there and always pass true to the super ctor?).

docvalues of what type? Maybe we can defer this one, since there is still some stuff to sort out about it (e.g. parent-child)

rjernst commented:
>> It is also worthwhile to note that with this PR, _uid will have doc values enabled (because it is indexed and not analyzed, although maybe we should be explicit there and always pass true to the super ctor?).
> docvalues of what type? Maybe we can defer this one, since there is still some stuff to sort out about it (e.g. parent-child)

Ok, I don't want to open up this can of worms. I'll set it to false explicitly, and we can enable it in future issue.

rjernst commented:
@rmuir @jpountz I pushed another commit addressing your comments.

jpountz commented:
LGTM
